common/src/main/java/com/wynntils/core/text/StyledText.java

```diff
@@ -0,0 +1,487 @@
+/*
+ * Copyright © Wynntils 2023.
+ * This file is released under AGPLv3. See LICENSE for full license details.
+ */
+package com.wynntils.core.text;
+
+import com.google.common.collect.Iterables;
+import com.wynntils.utils.mc.ComponentUtils;
+import com.wynntils.utils.type.Pair;
+import java.util.Collections;
+import java.util.Deque;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Objects;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+import net.minecraft.ChatFormatting;
+import net.minecraft.network.chat.ClickEvent;
+import net.minecraft.network.chat.Component;
+import net.minecraft.network.chat.HoverEvent;
+import net.minecraft.network.chat.MutableComponent;
+import net.minecraft.network.chat.Style;
+import org.apache.commons.lang3.ArrayUtils;
+
+public final class StyledText {
+    private final Component temporaryWorkaround;
+
+    private final List<StyledTextPart> parts;
+
+    private final List<ClickEvent> clickEvents;
+    private final List<HoverEvent> hoverEvents;
+
+    StyledText(Component component) {
+        temporaryWorkaround = component;
+
+        parts = new LinkedList<>();
+        clickEvents = new LinkedList<>();
+        hoverEvents = new LinkedList<>();
+
+        // Walk the component tree using DFS
+        // Save the style of the parent component so we can inherit it
+        Deque<Pair<Component, Style>> deque = new LinkedList<>();
+
+        deque.add(new Pair<>(component, Style.EMPTY));
+
+        while (!deque.isEmpty()) {
+            Pair<Component, Style> currentPair = deque.pop();
+            Component current = currentPair.key();
+            Style parentStyle = currentPair.value();
+
+            String componentString =
+                    MutableComponent.create(current.getContents()).getString();
+
+            StyledTextPart styledTextPart = StyledTextPart.fromComponentWithPossibleFormattingCodes(
+                    componentString, current.getStyle(), this, parentStyle);
+
+            List<Pair<Component, Style>> siblingPairs = current.getSiblings().stream()
```

Comment:
This looks odd, too. Are you trying to do here what I describe above? But current cannot have any siblings, right? You just created it from current.getContents()).getString()...

Reply:
Ah, now I understand. This is for your DFS traversal.